src/MixedQuoter.sol

```diff
@@ -0,0 +1,327 @@
+            V3SmartRouterHelper.getReserves(factoryV2, params.tokenIn, params.tokenOut);
+        amountOut = V3SmartRouterHelper.getAmountOut(params.amountIn, reserveIn, reserveOut);
+    }
+
+    /**
+     * Stable *************************************************
+     */
+
+    /// @dev Fetch an exactIn quote for a Stable pair on chain
+    function quoteExactInputSingleStable(QuoteExactInputSingleStableParams memory params)
+        public
+        view
+        override
+        returns (uint256 amountOut)
+    {
+        (uint256 i, uint256 j, address swapContract) =
+            V3SmartRouterHelper.getStableInfo(factoryStable, params.tokenIn, params.tokenOut, params.flag);
+        amountOut = IStableSwap(swapContract).get_dy(i, j, params.amountIn);
+    }
+
+    /**
+     * Mixed *************************************************
+     */
+    function quoteMixedExactInput(
+        address[] calldata paths,
+        bytes calldata actions,
+        bytes[] calldata params,
+        uint256 amountIn
+    ) external override returns (uint256 amountOut) {
+        uint256 numActions = actions.length;
+        if (numActions == 0) revert NoActions();
+        if (numActions != params.length || numActions != paths.length - 1) revert InputLengthMismatch();
+
+        for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+            address tokenIn = paths[actionIndex];
+            address tokenOut = paths[actionIndex + 1];
+            if (tokenIn == tokenOut) revert InvalidPath();
+
+            uint256 action = uint256(uint8(actions[actionIndex]));
+            if (action == MixedQuoterActions.V2_EXACT_INPUT_SINGLE) {
+                (tokenIn, tokenOut) = convertNativeToWETH(tokenIn, tokenOut);
+                // params[actionIndex] is zero bytes
+                amountIn = quoteExactInputSingleV2(
+                    QuoteExactInputSingleV2Params({tokenIn: tokenIn, tokenOut: tokenOut, amountIn: amountIn})
+                );
+            } else if (action == MixedQuoterActions.V3_EXACT_INPUT_SINGLE) {
+                (tokenIn, tokenOut) = convertNativeToWETH(tokenIn, tokenOut);
+                // params[actionIndex]: abi.encode(fee)
+                uint24 fee = abi.decode(params[actionIndex], (uint24));
+                (uint256 _amountOut,,,) = quoteExactInputSingleV3(
+                    QuoteExactInputSingleV3Params({
+                        tokenIn: tokenIn,
+                        tokenOut: tokenOut,
+                        amountIn: amountIn,
+                        fee: fee,
+                        sqrtPriceLimitX96: 0
+                    })
+                );
+                amountIn = _amountOut;
+            } else if (action == MixedQuoterActions.V4_CL_EXACT_INPUT_SINGLE) {
+                QuoteMixedV4ExactInputSingleParams memory clParams =
+                    abi.decode(params[actionIndex], (QuoteMixedV4ExactInputSingleParams));
+                (tokenIn, tokenOut) =
```

Comment:
curious why we need `isWethPool` when poolKey.currency0 == NATIVE woudl tell us if its native pool

Reply:
because the WETH will be converted to NATIVE by default.
If we want to quote v4 WETH pool , we need to tell the mixedQuoter do not convert WETH to NATIVE.
for example:
v3: token0-WETH
v4: WETH-token1

if no this flag , we can not tell mixedQuoter it is WETH-token1 or NATIVE-token1

Reply:
ok , good point.

Reply:
I just thought of one special case, so i added this, but it probably will not have. 😂
v4 pool: Native-WETH.
let me remove this.